### ISSUE TITLE
Future direction: Non-copyable `Span`

### ISSUE THREAD

**Issue by stevapple** [apple/swift-distributed-tracing#126]
Given that `Span` should be ended exactly once, and should not be shared between subprocesses or preserved after termination, it seems a potential client of the recently accepted [SE-0390](https://github.com/apple/swift-evolution/blob/main/proposals/0390-noncopyable-structs-and-enums.md). With Swift 5.9+ we can mark `Span` as non-copyable, and only allow the operation to “borrow” the span. `span.end()` will consume the span once. We can also add `span.end()` in `deinit` to automatically terminate the span.

Note that this is unfortunately **very breaking**. Haven’t thought about if there’s a way to land this safely in the 1.x period.

**Comment by ktoso** [comment]
Yeah it would be nice but I'm a bit worried about just doing it for 5.9+ for example...

Definitely worth keeping in mind so let's keep this issue open
